### PR TITLE
Expose the Enumeration APIs from the Unfold module

### PR DESCRIPTION
**Pull request by rnjtranjan**
https://github.com/composewell/streamly/issues/2351

### CODE REVIEW

File: core/src/Streamly/Data/Unfold.hs [hunk 1]
```diff
@@ -85,6 +85,37 @@ module Streamly.Data.Unfold
...
+    , enumerateFromThenSmallBounded
+    , enumerateFromToSmall
+    , enumerateFromThenToSmall
+
+    -- ** Enumerating 'Fractional' Types
+    -- | Enumeration of 'Num' specialized to 'Fractional' types.
+    , enumerateFromFractional
+    , enumerateFromThenFractional
+    , enumerateFromToFractional
+    , enumerateFromThenToFractional
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Unfold.hs:117]: Why are we exposing all these internal APIs? Please read the issue. Expose APIs similar to what is exposed in the Stream module.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Data/Unfold.hs [hunk 2]
```diff
@@ -85,6 +85,9 @@ module Streamly.Data.Unfold
     -- ** Nesting
     , many
 
+    -- ** Enumerating 'Num' Types
+    , enumerate
+    , enumerateTo
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Unfold.hs:90]: We have to expose the Enumerable type class and its members as well, that was the main thing to do in this PR.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Data/Unfold.hs [hunk 3]
```diff
@@ -226,3 +230,58 @@ import Streamly.Internal.Data.Unfold
 -- streams are written, it is easy to adapt a stream to an unfold. If you are
 -- writing an unfold you can convert it to stream for free using
 -- 'Stream.unfold'.
+
+-- | Unfolds @from@ generating a stream starting with the element
```
Comment (adithyaov) [inline core/src/Streamly/Data/Unfold.hs:234]: The API is called `enumerate`
The documentation needs to be changed!
Why is it named `enumerate` and not `enumerateFrom`?
Why not change the name to `enumerate` in the internal module itself?

File: core/src/Streamly/Data/Unfold.hs [hunk 4]
```diff
@@ -226,3 +230,58 @@ import Streamly.Internal.Data.Unfold
...
+  -- [1.1,2.1,3.1,4.1]
+  --
+  -- >>> Stream.fold Fold.toList $ Stream.unfold Unfold.enumerateFromTo (1.1, 4.6)
+  -- [1.1,2.1,3.1,4.1,5.1]
+  --
+  -- @
+  --
+{-# INLINE enumerateTo #-}
+enumerateTo :: (Monad m, Enumerable a) => Unfold m (a, a) a
+enumerateTo = enumerateFromTo
```
Comment (adithyaov) [inline core/src/Streamly/Data/Unfold.hs:287]: The same comments as earlier.

File: core/src/Streamly/Data/Unfold.hs (hunk 3 above)
Comment (rnjtranjan) [reply core/src/Streamly/Data/Unfold.hs]: The name "enumerate" API is used in Stream, so to be the matching name.

File: core/src/Streamly/Data/Unfold.hs [hunk 5]
```diff
@@ -226,3 +230,58 @@ import Streamly.Internal.Data.Unfold
 -- streams are written, it is easy to adapt a stream to an unfold. If you are
 -- writing an unfold you can convert it to stream for free using
 -- 'Stream.unfold'.
+
+-- | Enumerates @from@ generating a stream starting with the element
+  -- @from@, enumerating up to 'maxBound' when the type is 'Bounded' or
+  -- generating an infinite stream when the type is not 'Bounded'.
+  --
+  -- >>> import qualified Streamly.Data.Stream as Stream
+  -- >>> import qualified Streamly.Data.Unfold as Unfold
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Unfold.hs:239]: Any imports should be part of #include "DocTestDataUnfold.hs" .

File: core/src/Streamly/Data/Unfold.hs [hunk 6]
```diff
@@ -226,3 +230,58 @@ import Streamly.Internal.Data.Unfold
...
+  --
+  -- @
+  -- >>> Stream.fold Fold.toList $ Stream.take 4 $ Stream.unfold Unfold.enumerateFrom (0 :: Int)
+  -- [0,1,2,3]
+  --
+  -- @
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Unfold.hs:245]: doctests should not be inside `@` blocks.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Data/Unfold.hs [hunk 7]
```diff
@@ -226,3 +230,46 @@ import Streamly.Internal.Data.Unfold
...
+  -- For 'Fractional' types, enumeration is numerically stable. However, no
+  -- overflow or underflow checks are performed.
+  --
+  -- @
+  -- >>> Stream.fold Fold.toList $ Stream.take 4 $ Stream.unfold Unfold.enumerateFrom 1.1
+  -- [1.1,2.1,3.1,4.1]
+  --
+{-# INLINE enumerate #-}
+enumerate :: (Monad m, Enumerable a) => Unfold m a a
+enumerate = enumerateFrom
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Unfold.hs:252]: There is no point of this function if it is the same as enumerateFrom.

File: core/src/Streamly/Data/Unfold.hs [hunk 8]
```diff
@@ -226,3 +230,46 @@ import Streamly.Internal.Data.Unfold
...
+  -- @
+  -- >>> Stream.fold Fold.toList $ Stream.unfold Unfold.enumerateFromTo (1.1, 4)
+  -- [1.1,2.1,3.1,4.1]
+  --
+  -- >>> Stream.fold Fold.toList $ Stream.unfold Unfold.enumerateFromTo (1.1, 4.6)
+  -- [1.1,2.1,3.1,4.1,5.1]
+  --
+{-# INLINE enumerateTo #-}
+enumerateTo :: (Monad m, Enumerable a) => Unfold m (a, a) a
+enumerateTo = enumerateFromTo
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Unfold.hs:275]: There is no point of this function if this is same as enumerateFromTo.

**Review by harendra-kumar — COMMENTED** [review]
I am going to rework and commit this.

**Pull request merged**